implementation/src/main/java/io/smallrye/jwt/auth/AbstractBearerTokenExtractor.java

```diff
@@ -0,0 +1,107 @@
+/**
+ * Copyright 2019 Red Hat, Inc, and individual contributors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package io.smallrye.jwt.auth;
+
+import org.eclipse.microprofile.jwt.JsonWebToken;
+import org.jboss.logging.Logger;
+
+import io.smallrye.jwt.auth.principal.JWTAuthContextInfo;
+import io.smallrye.jwt.auth.principal.JWTCallerPrincipal;
+import io.smallrye.jwt.auth.principal.JWTCallerPrincipalFactory;
+import io.smallrye.jwt.auth.principal.ParseException;
+
+/**
+ * Common functionality for classes extracting Bearer tokens from HTTP request
+ * headers (including Cookie) and converting the token string to a
+ * {@link JsonWebToken}.
+ *
+ *
+ * @author Michael Edgar {@literal <michael@xlate.io>}
+ */
+public abstract class AbstractBearerTokenExtractor {
+
+    private static Logger logger = Logger.getLogger(AbstractBearerTokenExtractor.class);
+
+    private final JWTAuthContextInfo authContextInfo;
+
+    protected AbstractBearerTokenExtractor(JWTAuthContextInfo authContextInfo) {
+        this.authContextInfo = authContextInfo;
+    }
+
+    /**
+     * Find a JWT Bearer token in the request by referencing the configurations
+     * found in the {@link JWTAuthContextInfo}. The resulting token may be found
+     * in a cookie or another HTTP header, either explicitly configured or the
+     * default 'Authorization' header.
+     *
+     * @return a JWT Bearer token or null if not found
+     */
+    public String getBearerToken() {
+        final String tokenHeaderName = authContextInfo.getTokenHeader();
+        final String bearerValue;
+
+        if ("Cookie".equals(tokenHeaderName)) {
+            String tokenCookieName = authContextInfo.getTokenCookie();
+
+            if (tokenCookieName == null) {
+                tokenCookieName = "Bearer";
+            }
+
+            logger.debugf("tokenCookieName = %s", tokenCookieName);
+
+            bearerValue = getCookieValue(tokenCookieName);
+
+            if (bearerValue == null) {
+                logger.debugf("Cookie %s was null", tokenCookieName);
+            }
+        } else {
+            final String tokenHeader = getHeaderValue(tokenHeaderName);
+            logger.debugf("tokenHeaderName = %s", tokenHeaderName);
+
+            if (tokenHeader != null && tokenHeader.startsWith("Bearer ")) {
```

Comment:
@MikeEdgar Can you please change it to check for the 'Bearer' scheme only if it is 'Authorization' otherwise have the token value as is to support the case where the token is coming with some other header (this is the current code in Thornatl/Quarkus) and also do a case insensitive 'bearer' scheme check. It would be the last change I think... thanks

Reply:
@sberyozkin - done and pushed. Good catch.

Reply:
@MikeEdgar Thanks. Sure, please consider a follow up PR, may be a dedicated class for different categories of constants or HTTP only etc.

Your PR is looking perfect to me now :-), testing...